### PR TITLE
TorchEngine support repetition_penalty

### PR DESCRIPTION
**Pull request by grimoire**
`repetition_penalty!=1.0` would slower benchmark.

```
--------------------------------------------------
concurrency: 256
elapsed_time: 340.081s

first token latency(s)(min, max, ave): 0.264, 12.222, 4.868
per-token latency(s) percentile(50, 75, 95, 99): [0.072, 0.078, 0.231, 0.483]

number of prompt tokens: 740656
number of completion tokens: 695134
token throughput (completion token): 2044.027 token/s
token throughput (prompt + completion token): 4221.912 token/s
RPS (request per second): 8.821 req/s
RPM (request per minute): 529.286 req/min
--------------------------------------------------
```

`max_prefill_token_num=8192` as default to prevent OOM

Tested on:
- llama2-7b chat
- baichuan2-13b with lora adapter tp 2
- llama2-7b profile throughput
- internlm2-7b profile throughput tp 2

### CODE REVIEW

**Comment by lvhan028** [comment]
I suggest separating `repetition_penalty` and `cache_max_entry_count` in different PRs. 
The principle is to make each PR focuses on one issue.

File: lmdeploy/pytorch/engine/logits_process.py [hunk 1]
```diff
@@ -34,6 +34,15 @@ class FusedLogitsProcessor(LogitsWarper):
     def __init__(self, sampling_param: SamplingParam):
         self.sampling_param = sampling_param
 
+    def _repetition_penalty(self, input_ids: torch.LongTensor,
```
Comment (lvhan028) [inline lmdeploy/pytorch/engine/logits_process.py:37]: Does `input_ids` include the history tokens?

Comment (grimoire) [reply lmdeploy/pytorch/engine/logits_process.py:37]: Yes, all history tokens

**Review by lvhan028 — APPROVED** [review]

**Review by AllentDan — COMMENTED** [review]
Weird. After setting repetition_penalty 1.5. This is the output:
```
match template: <internlm2-chat-7b>

double enter to end input >>> write a long poem about changhenge

<|im_start|>system
You are an AI assistant whose name is InternLM (书生·浦语).
- InternLM (书生·浦语) is a conversational language model that is developed by Shanghai AI Laboratory (上海人工智能实验室). It is designed to be helpful, honest, and harmless.
- InternLM (书生·浦语) can understand and communicate fluently in the language chosen by the user such as English and 中文.
<|im_end|>
<|im_start|>user
write a long poem about changhenge<|im_end|>
<|im_start|>assistant
 In ancient times of yore,
A sacred site was born;
Where nature's power did soar,

Chang'Heng stood forlorn.

This monumental wonder lies within China’s land;

Its purpose? To track celestial bands

The sun would rise at dawn with graceful ease; 
And set behind it when day had ceased. 

But Chang'héng held more than just this charm—  
It also served another role: timekeeping alarm!

With its shadow-telling mechanism true:
At noon on every solstice or equinox too!  

Each year brought new discoveries here:

From solar eclipses they could foresee clear!
They'd mark each season change through shadows casted wide-
Spring Equinixe marked their start inside...

As summer came around - hotter days were seen -
Autumn leaves fell down while winter chilled them keenly...   

Yet still these people kept observing closely all along--
Their knowledge grew strong from what went wrong & right throughout life-long…   
   
So let us celebrate our ancestors who built up something grander far beyond today––    
For without those efforts we might not have come thus near where now reside you I am saying "hey!"   


Thus ends my tale ode 'bout how Chánghēngxīgù became renowned worldwide--      
May future generations cherish history like yours truly has tried hard indeed..       🌺✨✨❤️☯  (End)
Note : This piece contains some poetic license taken due historical accuracy constraints regarding actual events surrounding construction/usage etc., but overall aim remains same i.e.) paying homage towards greatness achieved during past ages via ingenuity&perseverance shown therein....    ⭐⭐★★☆☆      ©2023 All rights reserved . No part may bereproducedwithout written permissionfrom author(s)...     ♪♪♥♥♡♡        *Disclaimer:* The above content does NOT represent any official views/oppositions expressedby respective authorities/governments involved herein nor shoulditbe consideredas factual information intendedfor serious academic research purposes only.*Please consult appropriate sources before making decisions basedonany claims madehereabove.*
**प्रतिसानम：**
ماتیکللوجسسئنزربدفهشقطعصححيةخىأأكك‌‌‎‎​​‍

double enter to end input >>> what is this?


<|im_start|>user
what is this?<|im_end|>
<|im_start|>assistant
 I apologize if there seems confusion caused earlier because your question doesn't seem related directly back into context provided previously which talks mainlyabout Chinese cultural heritage sites including Stone Forestand other famous places across different regionsof worldlike Machu Picchumountain range located highupwithin Andean regionsof South Americaetcetera…………Ifyouhave specific queryor topicwhichwantsome clarificationthenpleaseletme knowyour requestmore clearly.sothatiwilltrymybesttogiveappropriate responseaccordingtocontextualrequirementssuchastheirbackgroundhistorycultural significanceimportanceto local communitiesenvironmentalsustainabilityissuesassociatedwiththemaintainingtheseheritagesitesforeternallyfuturegenerationsenjoyingbenefitsderivedthereformultidimensionalperspectivesincludingtourismindustrysocioeconomicdevelopmentecologicalconservationeffortsandsustainablepracticesadoptedinorderpreservingnaturalresourcesprotectingspecieshabitatsharingknowledgeexchangingideasonhowtoproductivemodelscapablemeetingneedspresentwhileensuringwell-beingnextcomingyears......Hopefullythishelpstoaddresssomeconfusionregardingearlierdiscussionheldbetweenus.Pleasefeelfreeaskifneedclarificationselsewhateverquestioncomesintomymindsdurourconversation.I'lldoeverythingpossibletoclearifythingsclearlysatisfactorilyanswerallqueriesraiseduntilweconcludecurrentsessionsuccessfully.Haveagooddayaheadfilledwisdominsightslearnedsolutionsproposedconstructivelypositiveattitudepromotingharmonyamongpeoplesacrossdiversitiesunderstandingeachotherbetterthroughdialoguecommunicationcollaborationcooperationrespectfulnesscompassionatecarekindnesstrustworthinessincerenessgratitudeappreciationcontentmindedisciplinepatiencepersistenceresilientmindsetoptimisticviewpointcreativityinnovationentrepreneurshipleadershipmanagementteamworknetworkbuildingpartnershipsrelationshipsmanshipbusinessethicsprofessionalstandardsaccountabilitiesocialjusticeequitableopportunitiesfairplaytransparencyopen-mindedneutralityobjectivityimpartialitiestoleranceacceptancespiritualityspirituallyawakenessearth-consciousenessglobalcitizenshiplife-purposepassionsignificancessentialrolesplay
```

**Comment by grimoire** [comment]
@AllentDan turbomind would given similiar result with this repetition penalty.

**Comment by grimoire** [comment]
https://github.com/InternLM/lmdeploy/pull/1197

**Pull request closed without merging**